# Multi-Resistance Antibiotic Susceptibility

## Fase 0 – Preparación y perfilado

In [1]:
# %% [markdown]
# ## Fase 0 – Preparación y perfilado del dataset

# %%
import pandas as pd
import numpy as np

# Opciones de visualización
pd.set_option("display.max_columns", 100)
pd.set_option("display.width", 140)

In [2]:
# %%
DATA_PATH = "Bacteria_dataset_Multiresictance.csv"

df = pd.read_csv(DATA_PATH)

df.shape, df.columns.tolist()

((10710, 27),
 ['ID',
  'Name',
  'Email',
  'Address',
  'age/gender',
  'Souches',
  'Diabetes',
  'Hypertension',
  'Hospital_before',
  'Infection_Freq',
  'AMX/AMP',
  'AMC',
  'CZ',
  'FOX',
  'CTX/CRO',
  'IPM',
  'GEN',
  'AN',
  'Acide nalidixique',
  'ofx',
  'CIP',
  'C',
  'Co-trimoxazole',
  'Furanes',
  'colistine',
  'Collection_Date',
  'Notes'])

In [3]:
df

,ID,Name,Email,Address,age/gender,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,Collection_Date,Notes
0,S290,Elizabeth Lawrence,elizabeth.lawrence@example.com,"6350 Robinson Loaf Apt. 447, Paulfurt, RI 30252",37/F,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,5 Fev 2025,Patient had fever??
1,S291,Tina Sanders,tina.sanders@example.com,"78594 Galloway Port Suite 762, South Tanyatown...",29/F,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,08/05/2021,تم العلاج
2,S292,Erin Cooke,erin.cooke@example.com,"76661 Isaiah Manors, North Benjamin, HI 85195",77/F,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,01/05/2022,NADA
3,S293,Mrs. Sharon Watson,mrs..sharon.watson@example.com,"1670 Thompson Way Apt. 520, Andrewbury, IL 11534",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2020-01-05,PACIENTO ENFERMO
4,S294,Erika Stewart,erika.stewart@example.com,"76838 Patrick Green Suite 538, Torresmouth, ME...",13/F,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,error,NADA
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10705,S10995,Mrs. Chelsea Brown MD,mrs..chelsea.brown.md@example.com,"23574 Palmer Lodge, South Aliciaton, GU 00728",17/M,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,2024-12-31,see file
10706,S10996,Mary Kelley,mary.kelley@example.com,"8194 Gonzalez Hill, Loveview, SD 31189",20/F,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,5 Fev 2025,Patient had fever??
10707,S10997,Alexandra Smith,alexandra.smith@example.com,"213 Yates Station, Port Meganstad, PW 17799",33/M,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,08/05/2021,Patient had fever??
10708,S10998,April Cox,april.cox@example.com,"985 Julia Freeway Apt. 753, Dianamouth, SD 07610",86/F,S10079 Escherichia coli,No,No,No,2.0,Intermediate,s,r,i,r,Intermediate,r,s,i,r,r,s,Intermediate,r,s,05/03/2022,see file


In [4]:
# %%
# Definir grupos lógicos de columnas (según el reporte y la metadata)

ID_COL = "ID"

PII_COLS = ["Name", "Email", "Address", "Notes"]  # se eliminarán más adelante

DEMOGRAPHIC_COLS = ["age/gender", "Souches"]

RISK_FACTOR_COLS = ["Diabetes", "Hypertension", "Hospital_before", "Infection_Freq"]

ANTIBIOTIC_COLS = [
    "AMX/AMP",
    "AMC",
    "CZ",
    "FOX",
    "CTX/CRO",
    "IPM",
    "GEN",
    "AN",
    "Acide nalidixique",
    "ofx",
    "CIP",
    "C",
    "Co-trimoxazole",
    "Furanes",
    "colistine",
]

META_COLS = ["Collection_Date"]

ALL_EXPECTED_COLS = (
    [ID_COL]
    + PII_COLS
    + DEMOGRAPHIC_COLS
    + RISK_FACTOR_COLS
    + ANTIBIOTIC_COLS
    + META_COLS
)

# Comprobemos que todo está en el DataFrame
missing_cols = [c for c in ALL_EXPECTED_COLS if c not in df.columns]
extra_cols = [c for c in df.columns if c not in ALL_EXPECTED_COLS]

missing_cols, extra_cols

([], [])

In [5]:
# %%
# Info básica de tipos, memoria y primeras filas
print("Shape:", df.shape)
print("\nTipos de datos:")
print(df.dtypes)

mem_mb = df.memory_usage(deep=True).sum() / 1024**2
print(f"\nUso aproximado de memoria: {mem_mb:.2f} MB")

Shape: (10710, 27)

Tipos de datos:
ID                   object
Name                 object
Email                object
Address              object
age/gender           object
Souches              object
Diabetes             object
Hypertension         object
Hospital_before      object
Infection_Freq       object
AMX/AMP              object
AMC                  object
CZ                   object
FOX                  object
CTX/CRO              object
IPM                  object
GEN                  object
AN                   object
Acide nalidixique    object
ofx                  object
CIP                  object
C                    object
Co-trimoxazole       object
Furanes              object
colistine            object
Collection_Date      object
Notes                object
dtype: object

Uso aproximado de memoria: 14.94 MB


In [6]:
# %%
def build_column_profile(data: pd.DataFrame, sample_size: int = 5) -> pd.DataFrame:
    """
    Construye un perfil inicial por columna:
    - dtype
    - n_distinct
    - n_missing
    - pct_missing
    - ejemplos de valores no nulos
    """
    records = []
    n_rows = len(data)

    for col in data.columns:
        series = data[col]
        n_missing = series.isna().sum()
        pct_missing = (n_missing / n_rows) * 100 if n_rows > 0 else np.nan
        n_distinct = series.nunique(dropna=True)

        sample_values = series.dropna().astype(str).unique()[:sample_size].tolist()

        records.append(
            {
                "columna": col,
                "dtype": str(series.dtype),
                "n_distinct": int(n_distinct),
                "n_missing": int(n_missing),
                "pct_missing": round(pct_missing, 2),
                "ejemplos": sample_values,
            }
        )

    profile_df = pd.DataFrame(records).sort_values("columna")
    return profile_df


column_profile = build_column_profile(df)
column_profile

,columna,dtype,n_distinct,n_missing,pct_missing,ejemplos
11,AMC,object,8,632,5.90,"[R, S, s, r, Intermediate]"
10,AMX/AMP,object,8,658,6.14,"[R, S, s, i, ?]"
17,AN,object,8,628,5.86,"[S, R, i, missing, s]"
18,Acide nalidixique,object,8,622,5.81,"[S, R, Intermediate, s, r]"
3,Address,object,10710,0,0.00,"[6350 Robinson Loaf Apt. 447, Paulfurt, RI 302..."
21,C,object,8,629,5.87,"[R, S, Intermediate, i, ?]"
20,CIP,object,8,633,5.91,"[S, R, s, Intermediate, i]"
14,CTX/CRO,object,8,624,5.83,"[R, S, r, i, ?]"
12,CZ,object,8,628,5.86,"[R, S, r, Intermediate, ?]"
22,Co-trimoxazole,object,8,640,5.98,"[S, R, Intermediate, s, r]"


In [7]:
# Si quieres ver solo algunas columnas (por ejemplo, factores de riesgo + antibióticos):
# %%
column_profile[column_profile["columna"].isin(RISK_FACTOR_COLS + ANTIBIOTIC_COLS)]

,columna,dtype,n_distinct,n_missing,pct_missing,ejemplos
11,AMC,object,8,632,5.90,"[R, S, s, r, Intermediate]"
10,AMX/AMP,object,8,658,6.14,"[R, S, s, i, ?]"
17,AN,object,8,628,5.86,"[S, R, i, missing, s]"
18,Acide nalidixique,object,8,622,5.81,"[S, R, Intermediate, s, r]"
21,C,object,8,629,5.87,"[R, S, Intermediate, i, ?]"
20,CIP,object,8,633,5.91,"[S, R, s, Intermediate, i]"
14,CTX/CRO,object,8,624,5.83,"[R, S, r, i, ?]"
12,CZ,object,8,628,5.86,"[R, S, r, Intermediate, ?]"
22,Co-trimoxazole,object,8,640,5.98,"[S, R, Intermediate, s, r]"
6,Diabetes,object,4,635,5.93,"[No, True, ?, missing]"


In [8]:
# Porcentaje de valores faltantes por columna
# %%
missing_summary = (
    df.isna()
    .mean()
    .mul(100)
    .round(2)
    .rename("pct_missing")
    .to_frame()
    .sort_values("pct_missing", ascending=False)
)

missing_summary

,pct_missing
Collection_Date,6.25
AMX/AMP,6.14
GEN,6.00
Souches,5.98
Co-trimoxazole,5.98
Infection_Freq,5.97
Hospital_before,5.95
Furanes,5.95
colistine,5.94
Diabetes,5.93


In [9]:
# %%
# Duplicados por ID
n_ids = df[ID_COL].nunique(dropna=True)
n_rows = len(df)
n_dup_ids = n_rows - n_ids

print(f"Número de filas: {n_rows}")
print(f"Número de IDs únicos ({ID_COL}): {n_ids}")
print(f"Filas con ID duplicado (potencialmente problemáticas): {n_dup_ids}")

# Duplicados de fila completa
n_dup_rows = df.duplicated().sum()
print(f"Filas completamente duplicadas: {n_dup_rows}")

Número de filas: 10710
Número de IDs únicos (ID): 10710
Filas con ID duplicado (potencialmente problemáticas): 0
Filas completamente duplicadas: 0


In [10]:
# 💊 7. Exploración de codificaciones S/I/R por antibiótico
# Aquí solo miramos los valores distintos y su frecuencia relativa, para ver el desorden que luego normalizaremos.
# %%
def value_counts_normalized(series: pd.Series, max_categories: int = 15) -> pd.DataFrame:
    """Devuelve una tabla con conteos y porcentaje relativo."""
    vc = (
        series.astype("string")
              .str.strip()
              .str.lower()
              .value_counts(dropna=False)
    )

    total = vc.sum()
    df_vc = (
        vc.to_frame(name="count")
          .reset_index()
          .rename(columns={"index": "valor"})
    )
    df_vc["pct"] = (df_vc["count"] / total * 100).round(2)
    return df_vc.head(max_categories)


for col in ANTIBIOTIC_COLS:
    print("=" * 80)
    print(f"Antibiótico: {col}")
    display(value_counts_normalized(df[col]))


Antibiótico: AMX/AMP


,AMX/AMP,count,pct
0,r,5712,53.33
1,s,4061,37.92
2,<NA>,658,6.14
3,i,93,0.87
4,intermediate,91,0.85
5,missing,51,0.48
6,?,44,0.41


Antibiótico: AMC


,AMC,count,pct
0,r,5815,54.3
1,s,3970,37.07
2,<NA>,632,5.9
3,i,97,0.91
4,intermediate,75,0.7
5,missing,66,0.62
6,?,55,0.51


Antibiótico: CZ


,CZ,count,pct
0,r,5690,53.13
1,s,4097,38.25
2,<NA>,628,5.86
3,intermediate,98,0.92
4,i,72,0.67
5,missing,66,0.62
6,?,59,0.55


Antibiótico: FOX


,FOX,count,pct
0,r,5735,53.55
1,s,4049,37.81
2,<NA>,625,5.84
3,i,93,0.87
4,intermediate,80,0.75
5,?,71,0.66
6,missing,57,0.53


Antibiótico: CTX/CRO


,CTX/CRO,count,pct
0,r,5748,53.67
1,s,4026,37.59
2,<NA>,624,5.83
3,i,95,0.89
4,intermediate,88,0.82
5,?,68,0.63
6,missing,61,0.57


Antibiótico: IPM


,IPM,count,pct
0,r,5717,53.38
1,s,4054,37.85
2,<NA>,633,5.91
3,i,94,0.88
4,intermediate,92,0.86
5,missing,62,0.58
6,?,58,0.54


Antibiótico: GEN


,GEN,count,pct
0,s,7854,73.33
1,r,1936,18.08
2,<NA>,643,6.0
3,i,84,0.78
4,intermediate,83,0.77
5,missing,56,0.52
6,?,54,0.5


Antibiótico: AN


,AN,count,pct
0,s,7875,73.53
1,r,1905,17.79
2,<NA>,628,5.86
3,intermediate,94,0.88
4,i,83,0.77
5,missing,66,0.62
6,?,59,0.55


Antibiótico: Acide nalidixique


,Acide nalidixique,count,pct
0,s,8379,78.24
1,r,1388,12.96
2,<NA>,622,5.81
3,intermediate,99,0.92
4,i,91,0.85
5,?,73,0.68
6,missing,58,0.54


Antibiótico: ofx


,ofx,count,pct
0,s,8406,78.49
1,r,1381,12.89
2,<NA>,618,5.77
3,i,88,0.82
4,intermediate,82,0.77
5,?,76,0.71
6,missing,59,0.55


Antibiótico: CIP


,CIP,count,pct
0,s,8331,77.79
1,r,1447,13.51
2,<NA>,633,5.91
3,intermediate,96,0.9
4,i,83,0.77
5,?,61,0.57
6,missing,59,0.55


Antibiótico: C


,C,count,pct
0,s,8390,78.34
1,r,1385,12.93
2,<NA>,629,5.87
3,i,97,0.91
4,intermediate,85,0.79
5,missing,66,0.62
6,?,58,0.54


Antibiótico: Co-trimoxazole


,Co-trimoxazole,count,pct
0,s,8370,78.15
1,r,1413,13.19
2,<NA>,640,5.98
3,intermediate,97,0.91
4,i,77,0.72
5,?,67,0.63
6,missing,46,0.43


Antibiótico: Furanes


,Furanes,count,pct
0,s,8457,78.96
1,r,1326,12.38
2,<NA>,637,5.95
3,i,93,0.87
4,intermediate,81,0.76
5,?,65,0.61
6,missing,51,0.48


Antibiótico: colistine


,colistine,count,pct
0,s,8428,78.69
1,r,1338,12.49
2,<NA>,636,5.94
3,intermediate,96,0.9
4,i,95,0.89
5,missing,65,0.61
6,?,52,0.49


In [11]:
# ❤️‍🩹 8. Exploración de codificaciones en factores de riesgo
# %%
for col in RISK_FACTOR_COLS:
    print("=" * 80)
    print(f"Factor de riesgo: {col}")
    display(value_counts_normalized(df[col]))

Factor de riesgo: Diabetes


,Diabetes,count,pct
0,no,7883,73.6
1,true,2064,19.27
2,<NA>,635,5.93
3,?,64,0.6
4,missing,64,0.6


Factor de riesgo: Hypertension


,Hypertension,count,pct
0,no,7470,69.75
1,yes,2477,23.13
2,<NA>,630,5.88
3,missing,73,0.68
4,?,60,0.56


Factor de riesgo: Hospital_before


,Hospital_before,count,pct
0,no,7008,65.43
1,yes,2939,27.44
2,<NA>,637,5.95
3,missing,64,0.6
4,?,62,0.58


Factor de riesgo: Infection_Freq


,Infection_Freq,count,pct
0,2.0,2966,27.69
1,1.0,2889,26.97
2,3.0,1963,18.33
3,0.0,1896,17.7
4,<NA>,639,5.97
5,missing,131,1.22
6,unknown,81,0.76
7,error,81,0.76
8,?,64,0.6


In [12]:
# 👤 9. Exploración rápida de age/gender y Souches
# %%
print("=== age/gender (ejemplos y distribución) ===")
display(value_counts_normalized(df["age/gender"], max_categories=20))

print("\n=== Souches (ejemplos de especies) ===")
species_sample = (
    df["Souches"]
    .dropna()
    .astype(str)
    .str.strip()
    .value_counts()
    .head(20)
    .to_frame(name="count")
)
species_sample

=== age/gender (ejemplos y distribución) ===


,age/gender,count,pct
0,<NA>,621,5.8
1,40/f,167,1.56
2,50/f,159,1.48
3,46/f,154,1.44
4,48/f,149,1.39
5,45/f,148,1.38
6,38/f,137,1.28
7,41/f,136,1.27
8,47/f,135,1.26
9,44/f,131,1.22



=== Souches (ejemplos de especies) ===


,count
Souches,
?,63
missing,60
S687 E. coli,3
S2424 Prot.eus mirabilis,3
S4989 Acinetobacter baumannii,3
S564 Escherichia coli,3
S4809 Escherichia coli,3
S4518 Escherichia coli,3
S4777 Escherichia coli,3


## Fase 1 — Limpieza y normalización

In [13]:
# %% [markdown]
# ## Fase 1 – Limpieza y normalización
# ### 1. Extracción del estado desde Address (columna State)

# %%
import re
import numpy as np

# Diccionario de estados de USA: nombre completo -> código de 2 letras
US_STATES = {
    "Alabama": "AL",
    "Alaska": "AK",
    "Arizona": "AZ",
    "Arkansas": "AR",
    "California": "CA",
    "Colorado": "CO",
    "Connecticut": "CT",
    "Delaware": "DE",
    "District of Columbia": "DC",
    "Florida": "FL",
    "Georgia": "GA",
    "Hawaii": "HI",
    "Idaho": "ID",
    "Illinois": "IL",
    "Indiana": "IN",
    "Iowa": "IA",
    "Kansas": "KS",
    "Kentucky": "KY",
    "Louisiana": "LA",
    "Maine": "ME",
    "Maryland": "MD",
    "Massachusetts": "MA",
    "Michigan": "MI",
    "Minnesota": "MN",
    "Mississippi": "MS",
    "Missouri": "MO",
    "Montana": "MT",
    "Nebraska": "NE",
    "Nevada": "NV",
    "New Hampshire": "NH",
    "New Jersey": "NJ",
    "New Mexico": "NM",
    "New York": "NY",
    "North Carolina": "NC",
    "North Dakota": "ND",
    "Ohio": "OH",
    "Oklahoma": "OK",
    "Oregon": "OR",
    "Pennsylvania": "PA",
    "Rhode Island": "RI",
    "South Carolina": "SC",
    "South Dakota": "SD",
    "Tennessee": "TN",
    "Texas": "TX",
    "Utah": "UT",
    "Vermont": "VT",
    "Virginia": "VA",
    "Washington": "WA",
    "West Virginia": "WV",
    "Wisconsin": "WI",
    "Wyoming": "WY",
}

# Patrón para capturar códigos de 2 letras de estados
STATE_ABBREVS_PATTERN = r"\b(" + "|".join(US_STATES.values()) + r")\b"


def extract_state_from_address(address: str) -> str | float:
    """
    Extrae el estado de USA desde un campo Address.
    Prioriza el código de 2 letras; si no lo encuentra,
    busca por nombre completo y devuelve el código.
    Si no encuentra nada, devuelve np.nan.
    """
    if not isinstance(address, str):
        return np.nan

    text = address.strip()
    if not text:
        return np.nan

    # 1. Buscar código de 2 letras
    match_abbrev = re.search(STATE_ABBREVS_PATTERN, text, flags=re.IGNORECASE)
    if match_abbrev:
        return match_abbrev.group(1).upper()

    # 2. Buscar por nombre completo
    lower_text = text.lower()
    for name, code in US_STATES.items():
        if re.search(r"\b" + re.escape(name) + r"\b", lower_text):
            return code

    # 3. Nada encontrado
    return np.nan

In [14]:
# %%
# Creamos una copia de trabajo para la limpieza
df_clean = df.copy()

# Crear la nueva columna State a partir de Address
df_clean["State"] = df_clean["Address"].apply(extract_state_from_address)

# Vista rápida de Address vs State para verificar
df_clean[["Address", "State"]].head(15)

,Address,State
0,"6350 Robinson Loaf Apt. 447, Paulfurt, RI 30252",RI
1,"78594 Galloway Port Suite 762, South Tanyatown...",HI
2,"76661 Isaiah Manors, North Benjamin, HI 85195",HI
3,"1670 Thompson Way Apt. 520, Andrewbury, IL 11534",IL
4,"76838 Patrick Green Suite 538, Torresmouth, ME...",ME
5,"USNS Neal, FPO AA 98816",NaN
6,"389 Ferguson Island, West Jefferymouth, MT 89882",MT
7,"256 Bartlett Fork Suite 334, East Sherri, MI 9...",MI
8,"77940 Richard Brook Apt. 748, South Jessica, G...",GA
9,"1979 Kayla Squares Suite 125, West Cynthia, MO...",MO


In [15]:
# %%
# Porcentaje de missing en la nueva columna State
n_rows = len(df_clean)
n_missing_state = df_clean["State"].isna().sum()
pct_missing_state = n_missing_state / n_rows * 100

print(f"Filas totales: {n_rows}")
print(
    f"Filas sin estado reconocido (State == NA): {n_missing_state} ({pct_missing_state:.2f}%)"
)

print("\nDistribución de State (top 20):")
display(
    df_clean["State"]
    .value_counts(dropna=False)
    .to_frame(name="count")
    .assign(pct=lambda s: (s["count"] / n_rows * 100).round(2))
    .head(20)
)

Filas totales: 10710
Filas sin estado reconocido (State == NA): 2372 (22.15%)

Distribución de State (top 20):


,count,pct
State,,
NaN,2372,22.15
WY,196,1.83
ND,187,1.75
IA,186,1.74
SD,181,1.69
AR,180,1.68
LA,179,1.67
NC,177,1.65
AK,174,1.62


In [16]:
# %%
# Rellenar NA en State con la categoría 'no georeferenciado'
df_clean["State"] = df_clean["State"].fillna("no georeferenciado")

# Verificar que ya no haya NA en State
n_rows = len(df_clean)
n_missing_state = df_clean["State"].isna().sum()
pct_missing_state = n_missing_state / n_rows * 100

print(f"Filas totales: {n_rows}")
print(
    f"Filas con State NA después del relleno: {n_missing_state} ({pct_missing_state:.2f}%)"
)

df_clean["State"].value_counts(dropna=False).head(20).to_frame("count").assign(
    pct=lambda s: (s["count"] / n_rows * 100).round(2)
)

Filas totales: 10710
Filas con State NA después del relleno: 0 (0.00%)


,count,pct
State,,
no georeferenciado,2372,22.15
WY,196,1.83
ND,187,1.75
IA,186,1.74
SD,181,1.69
AR,180,1.68
LA,179,1.67
NC,177,1.65
AK,174,1.62


In [17]:
# %%
# Eliminar Address, Name y Email después de extraer State (PII que no usaremos)
df_clean = df_clean.drop(columns=PII_COLS)

df_clean

,ID,age/gender,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,Collection_Date,State
0,S290,37/F,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,5 Fev 2025,RI
1,S291,29/F,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,08/05/2021,HI
2,S292,77/F,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,01/05/2022,HI
3,S293,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2020-01-05,IL
4,S294,13/F,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,error,ME
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10705,S10995,17/M,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,2024-12-31,no georeferenciado
10706,S10996,20/F,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,5 Fev 2025,SD
10707,S10997,33/M,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,08/05/2021,no georeferenciado
10708,S10998,86/F,S10079 Escherichia coli,No,No,No,2.0,Intermediate,s,r,i,r,Intermediate,r,s,i,r,r,s,Intermediate,r,s,05/03/2022,SD


In [18]:
# %% [markdown]
# ### Limpieza y estandarización de Collection_Date

# %%


def clean_collection_date(value: object) -> object:
    """
    Estandariza los valores de Collection_Date:
    - Convierte valores no informativos a NaN (error, missing, None, '?').
    - Corrige 'Fev' -> 'Feb' (portugués -> inglés).
    - Corrige el typo '3 Jan 1019' -> '3 Jan 2019'.
    - Deja el resto tal cual para que luego pandas.to_datetime los pueda parsear.
    """
    if not isinstance(value, str):
        return np.nan

    v = value.strip()
    if not v:
        return np.nan

    lower = v.lower()

    # Valores explícitamente no válidos / anómalos
    invalid_tokens = {"?", "none", "missing", "error"}
    if lower in invalid_tokens:
        return np.nan

    # Normalizar nombre de mes 'Fev' -> 'Feb'
    # (usado en el dataset para febrero en portugués)
    v = v.replace("Fev", "Feb")

    # Corregir typo específico del dataset: '3 Jan 1019' -> '3 Jan 2019'
    # Lo dejamos explícito para trazabilidad.
    if v == "3 Jan 1019":
        v = "3 Jan 2019"

    return v

In [19]:
# %%
# Aplicar limpieza sobre la columna original de fechas
df_clean["Collection_Date_std"] = df_clean["Collection_Date"].apply(
    clean_collection_date
)

# Vista rápida de antes vs después
df_clean[["Collection_Date", "Collection_Date_std"]]

,Collection_Date,Collection_Date_std
0,5 Fev 2025,5 Feb 2025
1,08/05/2021,08/05/2021
2,01/05/2022,01/05/2022
3,2020-01-05,2020-01-05
4,error,NaN
...,...,...
10705,2024-12-31,2024-12-31
10706,5 Fev 2025,5 Feb 2025
10707,08/05/2021,08/05/2021
10708,05/03/2022,05/03/2022


In [20]:
# %%
# Convertir la columna estandarizada a datetime
# dayfirst=True porque hay fechas en formato dd/mm/yyyy
df_clean["Collection_Date_dt"] = pd.to_datetime(
    df_clean["Collection_Date_std"],
    errors="coerce",
    dayfirst=True,
)

# Verificar el resultado
print("Resumen de Collection_Date_dt:")
print(df_clean["Collection_Date_dt"].describe())

print("\nValores no parseables (NaT) tras la conversión:")
n_rows = len(df_clean)
n_nat = df_clean["Collection_Date_dt"].isna().sum()
print(f"NaT: {n_nat} ({n_nat / n_rows * 100:.2f}%)")

Resumen de Collection_Date_dt:
count                             1396
mean     2022-02-10 19:31:48.309455616
min                2019-01-03 00:00:00
25%                2019-01-03 00:00:00
50%                2025-02-05 00:00:00
75%                2025-02-05 00:00:00
max                2025-02-05 00:00:00
Name: Collection_Date_dt, dtype: object

Valores no parseables (NaT) tras la conversión:
NaT: 9314 (86.97%)


In [21]:
df_clean

,ID,age/gender,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,Collection_Date,State,Collection_Date_std,Collection_Date_dt
0,S290,37/F,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,5 Fev 2025,RI,5 Feb 2025,2025-02-05
1,S291,29/F,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,08/05/2021,HI,08/05/2021,NaT
2,S292,77/F,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,01/05/2022,HI,01/05/2022,NaT
3,S293,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2020-01-05,IL,2020-01-05,NaT
4,S294,13/F,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,error,ME,NaN,NaT
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10705,S10995,17/M,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,2024-12-31,no georeferenciado,2024-12-31,NaT
10706,S10996,20/F,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,5 Fev 2025,SD,5 Feb 2025,2025-02-05
10707,S10997,33/M,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,08/05/2021,no georeferenciado,08/05/2021,NaT
10708,S10998,86/F,S10079 Escherichia coli,No,No,No,2.0,Intermediate,s,r,i,r,Intermediate,r,s,i,r,r,s,Intermediate,r,s,05/03/2022,SD,05/03/2022,NaT


In [22]:
# %%
# Distribución por año y mes para ver que todo quedó razonable
print("Conteo por año de Collection_Date_dt:")
display(
    df_clean["Collection_Date_dt"]
    .dt.year.value_counts(dropna=False)
    .sort_index()
    .to_frame(name="count")
)

print("\nConteo por mes (1-12):")
display(
    df_clean["Collection_Date_dt"]
    .dt.month.value_counts(dropna=False)
    .sort_index()
    .to_frame(name="count")
)

Conteo por año de Collection_Date_dt:


,count
Collection_Date_dt,
2019.0,684
2025.0,712
NaN,9314



Conteo por mes (1-12):


,count
Collection_Date_dt,
1.0,684
2.0,712
NaN,9314


In [23]:
# Remove collection_date and Collection_Date_std columns
df_clean = df_clean.drop(columns=["Collection_Date", "Collection_Date_std"])
df_clean

,ID,age/gender,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt
0,S290,37/F,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05
1,S291,29/F,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT
2,S292,77/F,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT
3,S293,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,IL,NaT
4,S294,13/F,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10705,S10995,17/M,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT
10706,S10996,20/F,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05
10707,S10997,33/M,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT
10708,S10998,86/F,S10079 Escherichia coli,No,No,No,2.0,Intermediate,s,r,i,r,Intermediate,r,s,i,r,r,s,Intermediate,r,s,SD,NaT


In [24]:
# %% [markdown]
# ### Limpieza y división de la columna age/gender → Age_years y Gender

# %%

INVALID_AGE_GENDER_TOKENS = {"?", "none", "missing", "error", "unknown", "na", "n/a"}


def parse_age_gender(value: object):
    """
    Recibe un valor de la columna 'age/gender' y devuelve:
    (Age_years, Gender)

    - Age_years: número (int/float) o np.nan
    - Gender: 'M', 'F' o np.nan
    """
    if not isinstance(value, str):
        return np.nan, np.nan

    v = value.strip()
    if not v:
        return np.nan, np.nan

    lower = v.lower()
    if lower in INVALID_AGE_GENDER_TOKENS:
        return np.nan, np.nan

    # Intentar separar por '/'
    # Ejemplos: "37/F", "57/M", "25/", "??/M"
    parts = v.split("/")
    age_part = parts[0].strip() if len(parts) >= 1 else ""
    gender_part = parts[1].strip() if len(parts) >= 2 else ""

    # ---- Parseo de Edad ----
    age_val = np.nan
    if age_part:
        # Buscar un número de 1 a 3 dígitos dentro del texto
        m = re.search(r"\d{1,3}", age_part)
        if m:
            age_num = int(m.group(0))
            # Filtrar edades absurdas
            if 0 <= age_num <= 120:
                age_val = float(age_num)

    # ---- Parseo de Género ----
    gender_val = np.nan

    # Normalizar el texto de género
    g_text = gender_part.lower()

    # Si no viene nada a la derecha del '/', intentar buscar género en el string completo
    if not g_text:
        g_text = lower

    # Mapeos típicos a M/F (incluye variantes en inglés/español)
    if any(x in g_text for x in ["m", "male", "masculino", "hombre"]) and not any(
        x in g_text for x in ["f", "female", "femenino", "mujer"]
    ):
        gender_val = "Male"
    elif any(x in g_text for x in ["f", "female", "femenino", "mujer"]) and not any(
        x in g_text for x in ["m", "male", "masculino", "hombre"]
    ):
        gender_val = "Female"
    else:
        # Caso ambiguo o irreconocible → np.nan
        gender_val = np.nan

    return age_val, gender_val

In [25]:
# %%
# Aplicar parseo a toda la columna age/gender
parsed_age_gender = df_clean["age/gender"].apply(parse_age_gender)

# Desempaquetar la tupla en dos nuevas columnas
df_clean["Age_years"] = parsed_age_gender.apply(lambda x: x[0])
df_clean["Gender"] = parsed_age_gender.apply(lambda x: x[1])

# Vista rápida de la transformación
df_clean[["age/gender", "Age_years", "Gender"]]

,age/gender,Age_years,Gender
0,37/F,37.0,Female
1,29/F,29.0,Female
2,77/F,77.0,Female
3,NaN,NaN,NaN
4,13/F,13.0,Female
...,...,...,...
10705,17/M,17.0,Male
10706,20/F,20.0,Female
10707,33/M,33.0,Male
10708,86/F,86.0,Female


In [26]:
# %%
n_rows = len(df_clean)

print("=== Age_years ===")
print(df_clean["Age_years"].describe())

n_missing_age = df_clean["Age_years"].isna().sum()
print(f"\nMissing en Age_years: {n_missing_age} ({n_missing_age / n_rows * 100:.2f}%)")

print("\n=== Gender ===")
print(
    df_clean["Gender"]
    .value_counts(dropna=False)
    .to_frame("count")
    .assign(pct=lambda s: (s["count"] / n_rows * 100).round(2))
)

n_missing_gender = df_clean["Gender"].isna().sum()
print(
    f"\nMissing en Gender: {n_missing_gender} ({n_missing_gender / n_rows * 100:.2f}%)"
)

=== Age_years ===
count    9947.000000
mean       45.601789
std        24.891645
min         0.000000
25%        25.000000
50%        45.000000
75%        67.000000
max        90.000000
Name: Age_years, dtype: float64

Missing en Age_years: 763 (7.12%)

=== Gender ===
        count    pct
Gender              
Female   7929  74.03
Male     2018  18.84
NaN       763   7.12

Missing en Gender: 763 (7.12%)


In [27]:
# Remove age/gender column
df_clean = df_clean.drop(columns=["age/gender"])
df_clean

,ID,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt,Age_years,Gender
0,S290,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05,37.0,Female
1,S291,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT,29.0,Female
2,S292,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT,77.0,Female
3,S293,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,IL,NaT,NaN,NaN
4,S294,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT,13.0,Female
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10705,S10995,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT,17.0,Male
10706,S10996,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05,20.0,Female
10707,S10997,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT,33.0,Male
10708,S10998,S10079 Escherichia coli,No,No,No,2.0,Intermediate,s,r,i,r,Intermediate,r,s,i,r,r,s,Intermediate,r,s,SD,NaT,86.0,Female


In [28]:
# %%
DEMOGRAPHIC_COLS = ["Age_years", "Gender", "Souches", "State"]

DEMOGRAPHIC_COLS

['Age_years', 'Gender', 'Souches', 'State']

In [29]:
# %% [markdown]
# ### Eliminación de filas sin información útil (ej. S293)

# %%
import pandas as pd

NON_INFO_TOKENS = {"?", "none", "missing", "error", "unknown", "na", "n/a", ""}


def is_non_informative(value: object) -> bool:
    """
    Devuelve True si el valor se considera 'sin información':
    - NaN
    - tokens como ?, missing, error, unknown...
    - 'no georeferenciado' (en State)
    """
    if pd.isna(value):
        return True

    s = str(value).strip().lower()
    if s in NON_INFO_TOKENS:
        return True

    # Para State, 'no georeferenciado' lo tratamos como sin info
    if s == "no georeferenciado":
        return True

    return False

In [30]:
# %%
# Columnas analíticas mínimas para considerar una fila "con contenido"
ANALYTIC_COLS_FOR_EMPTY = (
    [
        "Age_years",
        "Gender",
        "Souches",
        "State",
        "Collection_Date_dt",
        "Infection_Freq",
    ]
    + RISK_FACTOR_COLS  # Diabetes, Hypertension, Hospital_before, Infection_Freq
    + ANTIBIOTIC_COLS  # los 15 antibióticos
)

# Verificamos que todas existan en df_clean (por seguridad)
missing_for_empty = [c for c in ANALYTIC_COLS_FOR_EMPTY if c not in df_clean.columns]
if missing_for_empty:
    print("Ojo, faltan estas columnas en df_clean:", missing_for_empty)

In [31]:
# %%
# Aplicar is_non_informative a las columnas analíticas
mask_empty_rows = (
    df_clean[ANALYTIC_COLS_FOR_EMPTY].applymap(is_non_informative).all(axis=1)
)

n_empty = mask_empty_rows.sum()
n_rows = len(df_clean)

print(
    f"Filas sin información útil (solo ruido): {n_empty} de {n_rows} ({n_empty / n_rows * 100:.2f}%)"
)

C:\Users\samaf\AppData\Local\Temp\ipykernel_18804\3800851482.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  df_clean[ANALYTIC_COLS_FOR_EMPTY].applymap(is_non_informative).all(axis=1)


Filas sin información útil (solo ruido): 136 de 10710 (1.27%)


In [32]:
mask_empty_rows

0        False
1        False
2        False
3        False
4        False
         ...  
10705    False
10706    False
10707    False
10708    False
10709    False
Length: 10710, dtype: bool

In [33]:
# %%
# Ver algunos ejemplos de filas marcadas como vacías
df_clean.loc[mask_empty_rows, ["ID"] + ANALYTIC_COLS_FOR_EMPTY]

,ID,Age_years,Gender,Souches,State,Collection_Date_dt,Infection_Freq,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine
56,S346,NaN,NaN,NaN,no georeferenciado,NaT,NaN,NaN,?,NaN,NaN,NaN,NaN,?,?,?,NaN,NaN,missing,NaN,NaN,NaN,?,NaN,NaN,missing
92,S382,NaN,NaN,NaN,no georeferenciado,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
108,S398,NaN,NaN,NaN,no georeferenciado,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
109,S399,NaN,NaN,NaN,no georeferenciado,NaT,NaN,NaN,missing,?,NaN,?,?,NaN,missing,NaN,NaN,?,NaN,NaN,NaN,NaN,missing,NaN,NaN,NaN
311,S601,NaN,NaN,NaN,no georeferenciado,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10253,S10543,NaN,NaN,NaN,no georeferenciado,NaT,NaN,missing,?,NaN,NaN,NaN,missing,?,?,NaN,NaN,missing,NaN,NaN,missing,NaN,?,?,NaN,?
10343,S10633,NaN,NaN,NaN,no georeferenciado,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
10368,S10658,NaN,NaN,missing,no georeferenciado,NaT,NaN,NaN,missing,NaN,NaN,?,NaN,NaN,NaN,missing,NaN,?,?,missing,NaN,NaN,?,?,?,missing
10461,S10751,NaN,NaN,?,no georeferenciado,NaT,?,missing,NaN,?,?,NaN,NaN,NaN,?,missing,NaN,NaN,missing,?,missing,?,missing,?,missing,?


In [34]:
# %%
# Guardamos cuántas filas vamos a eliminar para registrarlo en la bitácora
rows_before = len(df_clean)
df_clean = df_clean.loc[~mask_empty_rows].reset_index(drop=True)
rows_after = len(df_clean)

print(f"Filas antes:  {rows_before}")
print(f"Filas después: {rows_after}")
print(f"Filas eliminadas por no contener información útil: {rows_before - rows_after}")

Filas antes:  10710
Filas después: 10574
Filas eliminadas por no contener información útil: 136


In [35]:
df_clean

,ID,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt,Age_years,Gender
0,S290,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05,37.0,Female
1,S291,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT,29.0,Female
2,S292,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT,77.0,Female
3,S293,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,IL,NaT,NaN,NaN
4,S294,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT,13.0,Female
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
10569,S10995,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT,17.0,Male
10570,S10996,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05,20.0,Female
10571,S10997,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT,33.0,Male
10572,S10998,S10079 Escherichia coli,No,No,No,2.0,Intermediate,s,r,i,r,Intermediate,r,s,i,r,r,s,Intermediate,r,s,SD,NaT,86.0,Female


In [36]:
# Removamos del dataset las filas que posean un Souches no informativo
rows_before = len(df_clean)
df_clean = df_clean.loc[~df_clean["Souches"].apply(is_non_informative)].reset_index(drop=True)
rows_after = len(df_clean)

print(f"Filas antes:  {rows_before}")
print(f"Filas después: {rows_after}")
print(f"Filas eliminadas por no contener información útil: {rows_before - rows_after}")
df_clean

Filas antes:  10574
Filas después: 9947
Filas eliminadas por no contener información útil: 627


,ID,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt,Age_years,Gender
0,S290,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05,37.0,Female
1,S291,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT,29.0,Female
2,S292,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT,77.0,Female
3,S294,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT,13.0,Female
4,S295,S295 Escherichia coli,True,No,No,3.0,R,S,R,R,S,S,S,R,S,S,S,S,S,S,S,no georeferenciado,NaT,57.0,Male
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9942,S10994,S8862 Escherichia coli,No,No,No,0.0,S,R,R,R,S,S,S,S,R,S,S,R,S,S,S,RI,NaT,43.0,Female
9943,S10995,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT,17.0,Male
9944,S10996,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05,20.0,Female
9945,S10997,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT,33.0,Male


In [37]:
# %% [markdown]
# ### Normalización de antibióticos a {S, I, R, NA} y creación de Rbin

# %%
# Tokens que significan "no dato" / no informativo (NA)
ANTIBIOTIC_NA_TOKENS = {
    "",
    "?",
    "none",
    "no test",
    "not done",
    "missing",
    "error",
    "unknown",
    "na",
    "n/a",
    "nan",
    "-",
    ".",
    "null",
}


def normalize_sir_token(raw: object) -> str | float:
    """
    Normaliza un token crudo del antibiograma a {S, I, R, np.nan}.
    """
    if pd.isna(raw):
        return np.nan

    s = str(raw).strip()
    if not s:
        return np.nan

    lower = s.casefold()  # casefold = lower robusto

    # 1) Tokens claramente NA
    if lower in ANTIBIOTIC_NA_TOKENS:
        return np.nan

    # 2) Normalización explícita
    #   - S: susceptible / sensible
    if lower in {"s", "sus", "susc", "susceptible", "sensitive", "sensible"}:
        return "S"

    #   - I: intermedio
    if lower in {"i", "int", "interm", "intermediate", "intermedio", "indeterminate"}:
        return "I"

    #   - R: resistente
    if lower in {"r", "res", "resistant", "resistance", "resistente"}:
        return "R"

    # 3) Heurísticas suaves por prefijo / substring
    if lower.startswith("susc") or " suscept" in lower or "sensible" in lower:
        return "S"

    if lower.startswith("interm") or " intermediate" in lower or " intermedio" in lower:
        return "I"

    if lower.startswith("resist") or " resistant" in lower or " resistente" in lower:
        return "R"

    # 4) Si no cae en ningún caso → lo marcamos como NA
    # (y lo registraremos como token anómalo en el resumen)
    return np.nan

In [38]:
# %%
sir_summary_records = []
unknown_tokens_by_col: dict[str, set[str]] = {}

for col in ANTIBIOTIC_COLS:
    raw_series = df_clean[col]

    # Normalizar a S/I/R/NA
    col_sir = f"{col}_SIR"
    df_clean[col_sir] = raw_series.apply(normalize_sir_token)

    # Crear Rbin: 1 si R, 0 si S o I, NaN si NA
    col_rbin = f"{col}_Rbin"
    df_clean[col_rbin] = (
        df_clean[col_sir].map({"R": 1, "S": 0, "I": 0}).astype("float")
    )  # float para permitir NaN

    # Resumen de distribución
    total_n = len(df_clean)
    counts = df_clean[col_sir].value_counts(dropna=False)

    n_S = counts.get("S", 0)
    n_I = counts.get("I", 0)
    n_R = counts.get("R", 0)
    n_NA = counts.get(np.nan, 0)

    sir_summary_records.append(
        {
            "antibiotico": col,
            "n_total": total_n,
            "n_S": int(n_S),
            "n_I": int(n_I),
            "n_R": int(n_R),
            "n_NA": int(n_NA),
            "pct_NA": round(n_NA / total_n * 100, 2),
        }
    )

    # Registrar tokens crudos que quedaron como NA pero no son NA-tokens explícitos
    # (sirven para revisar posibles anomalías de codificación)
    normalized = raw_series.apply(normalize_sir_token)
    mask_na = normalized.isna()

    raw_tokens_na = (
        raw_series[mask_na]
        .astype("string")
        .str.strip()
        .str.casefold()
        .dropna()
        .unique()
        .tolist()
    )

    # Filtrar los que ya sabemos que son NA "esperado"
    unknown_tokens = {
        tok for tok in raw_tokens_na if tok not in ANTIBIOTIC_NA_TOKENS and tok != "nan"
    }

    unknown_tokens_by_col[col] = unknown_tokens

sir_summary = pd.DataFrame(sir_summary_records)
sir_summary

,antibiotico,n_total,n_S,n_I,n_R,n_NA,pct_NA
0,AMX/AMP,9947,4056,182,5709,0,0.0
1,AMC,9947,3970,167,5810,0,0.0
2,CZ,9947,4093,165,5689,0,0.0
3,FOX,9947,4047,166,5734,0,0.0
4,CTX/CRO,9947,4021,180,5746,0,0.0
5,IPM,9947,4052,179,5716,0,0.0
6,GEN,9947,7853,163,1931,0,0.0
7,AN,9947,7875,168,1904,0,0.0
8,Acide nalidixique,9947,8375,186,1386,0,0.0
9,ofx,9947,8404,165,1378,0,0.0


In [39]:
# %%
# Mostrar tokens crudos no reconocidos por columna (para auditoría)
for col, tokens in unknown_tokens_by_col.items():
    print("=" * 80)
    print(f"Antibiótico: {col}")
    if tokens:
        print("Tokens crudos no reconocidos (mapeados a NA):")
        print(sorted(tokens))
    else:
        print("Sin tokens anómalos adicionales (solo NA esperados).")

Antibiótico: AMX/AMP
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: AMC
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: CZ
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: FOX
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: CTX/CRO
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: IPM
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: GEN
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: AN
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: Acide nalidixique
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: ofx
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: CIP
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: C
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: Co-trimoxazole
Sin tokens anómalos adicionales (solo NA esperados).
Antibiótico: Furanes
Sin tokens anómalos adicionales (solo NA

In [40]:
# %%
# Vista rápida de algunas columnas con sus versiones SIR y Rbin
cols_to_show = []
for base in ["AMX/AMP", "CIP", "colistine"]:
    cols_to_show.extend([base, f"{base}_SIR", f"{base}_Rbin"])

df_clean[cols_to_show]

,AMX/AMP,AMX/AMP_SIR,AMX/AMP_Rbin,CIP,CIP_SIR,CIP_Rbin,colistine,colistine_SIR,colistine_Rbin
0,R,R,1.0,S,S,0.0,S,S,0.0
1,S,S,0.0,S,S,0.0,S,S,0.0
2,S,S,0.0,R,R,1.0,S,S,0.0
3,S,S,0.0,S,S,0.0,S,S,0.0
4,R,R,1.0,S,S,0.0,S,S,0.0
...,...,...,...,...,...,...,...,...,...
9942,S,S,0.0,S,S,0.0,S,S,0.0
9943,S,S,0.0,S,S,0.0,S,S,0.0
9944,S,S,0.0,S,S,0.0,S,S,0.0
9945,R,R,1.0,S,S,0.0,S,S,0.0


In [41]:
df_clean

,ID,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt,Age_years,Gender,AMX/AMP_SIR,AMX/AMP_Rbin,AMC_SIR,AMC_Rbin,CZ_SIR,CZ_Rbin,FOX_SIR,FOX_Rbin,CTX/CRO_SIR,CTX/CRO_Rbin,IPM_SIR,IPM_Rbin,GEN_SIR,GEN_Rbin,AN_SIR,AN_Rbin,Acide nalidixique_SIR,Acide nalidixique_Rbin,ofx_SIR,ofx_Rbin,CIP_SIR,CIP_Rbin,C_SIR,C_Rbin,Co-trimoxazole_SIR,Co-trimoxazole_Rbin,Furanes_SIR,Furanes_Rbin,colistine_SIR,colistine_Rbin
0,S290,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05,37.0,Female,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0
1,S291,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT,29.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0
2,S292,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT,77.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0
3,S294,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT,13.0,Female,S,0.0,S,0.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0
4,S295,S295 Escherichia coli,True,No,No,3.0,R,S,R,R,S,S,S,R,S,S,S,S,S,S,S,no georeferenciado,NaT,57.0,Male,R,1.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9942,S10994,S8862 Escherichia coli,No,No,No,0.0,S,R,R,R,S,S,S,S,R,S,S,R,S,S,S,RI,NaT,43.0,Female,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0
9943,S10995,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT,17.0,Male,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0
9944,S10996,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05,20.0,Female,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0
9945,S10997,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT,33.0,Male,R,1.0,R,1.0,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0


In [42]:
# %% [markdown]
# ### Normalización de booleanos (Diabetes, Hypertension, Hospital_before) → {0,1,NA}

# %%
import numpy as np
import pandas as pd

BOOL_TRUE = {"yes", "y", "1", "true"}
BOOL_FALSE = {"no", "n", "0", "false"}


def normalize_bool_token(raw: object):
    """
    Normaliza un token booleano a {1.0, 0.0, NaN} siguiendo la regla:
    {yes, y, 1, true}  -> 1
    {no, n, 0, false}  -> 0
    resto              -> NaN
    """
    if pd.isna(raw):
        return np.nan

    s = str(raw).strip().casefold()
    if not s:
        return np.nan

    if s in BOOL_TRUE:
        return 1.0
    if s in BOOL_FALSE:
        return 0.0

    return np.nan


BOOLEAN_COLS = ["Diabetes", "Hypertension", "Hospital_before"]

for col in BOOLEAN_COLS:
    new_col = f"{col}_bin"
    df_clean[new_col] = df_clean[col].apply(normalize_bool_token)

    print("=" * 80)
    print(f"Columna: {col} → {new_col}")
    display(df_clean[[col, new_col]].head(10))

    # Resumen rápido
    n = len(df_clean)
    counts = df_clean[new_col].value_counts(dropna=False)
    print(
        counts.to_frame("count").assign(pct=lambda s: (s["count"] / n * 100).round(2))
    )

Columna: Diabetes → Diabetes_bin


,Diabetes,Diabetes_bin
0,No,0.0
1,True,1.0
2,True,1.0
3,No,0.0
4,True,1.0
5,No,0.0
6,No,0.0
7,No,0.0
8,No,0.0
9,True,1.0


              count    pct
Diabetes_bin              
0.0            7883  79.25
1.0            2064  20.75
Columna: Hypertension → Hypertension_bin


,Hypertension,Hypertension_bin
0,No,0.0
1,No,0.0
2,No,0.0
3,Yes,1.0
4,No,0.0
5,No,0.0
6,No,0.0
7,No,0.0
8,No,0.0
9,No,0.0


                  count   pct
Hypertension_bin             
0.0                7470  75.1
1.0                2477  24.9
Columna: Hospital_before → Hospital_before_bin


,Hospital_before,Hospital_before_bin
0,No,0.0
1,No,0.0
2,No,0.0
3,No,0.0
4,No,0.0
5,No,0.0
6,No,0.0
7,No,0.0
8,No,0.0
9,Yes,1.0


                     count    pct
Hospital_before_bin              
0.0                   7008  70.45
1.0                   2939  29.55


In [43]:
df_clean

,ID,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt,Age_years,Gender,AMX/AMP_SIR,AMX/AMP_Rbin,AMC_SIR,AMC_Rbin,CZ_SIR,CZ_Rbin,FOX_SIR,FOX_Rbin,CTX/CRO_SIR,CTX/CRO_Rbin,IPM_SIR,IPM_Rbin,GEN_SIR,GEN_Rbin,AN_SIR,AN_Rbin,Acide nalidixique_SIR,Acide nalidixique_Rbin,ofx_SIR,ofx_Rbin,CIP_SIR,CIP_Rbin,C_SIR,C_Rbin,Co-trimoxazole_SIR,Co-trimoxazole_Rbin,Furanes_SIR,Furanes_Rbin,colistine_SIR,colistine_Rbin,Diabetes_bin,Hypertension_bin,Hospital_before_bin
0,S290,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05,37.0,Female,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0
1,S291,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT,29.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0
2,S292,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT,77.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0
3,S294,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT,13.0,Female,S,0.0,S,0.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,0.0,1.0,0.0
4,S295,S295 Escherichia coli,True,No,No,3.0,R,S,R,R,S,S,S,R,S,S,S,S,S,S,S,no georeferenciado,NaT,57.0,Male,R,1.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9942,S10994,S8862 Escherichia coli,No,No,No,0.0,S,R,R,R,S,S,S,S,R,S,S,R,S,S,S,RI,NaT,43.0,Female,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0
9943,S10995,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT,17.0,Male,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0
9944,S10996,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05,20.0,Female,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,1.0
9945,S10997,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT,33.0,Male,R,1.0,R,1.0,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0


In [44]:
# %% [markdown]
# ### Normalización de especie (Souches) y tabla de alias

# %%
SOUCHES_NA_TOKENS = {"", "?", "none", "missing", "error", "unknown", "na", "n/a"}


def is_non_informative_souche(value: object) -> bool:
    """
    Marca si Souches no aporta información:
    - NaN
    - tokens tipo ?, missing, error, unknown, etc.
    """
    if pd.isna(value):
        return True

    s = str(value).strip().casefold()
    if not s:
        return True

    if s in SOUCHES_NA_TOKENS:
        return True

    return False


mask_souche_empty = df_clean["Souches"].apply(is_non_informative_souche)

n_empty_souche = mask_souche_empty.sum()
n_rows = len(df_clean)
print(
    f"Filas sin especie informativa (Souches vacío): {n_empty_souche} de {n_rows} ({n_empty_souche / n_rows * 100:.2f}%)"
)

# Inspección rápida de algunos casos antes de eliminar
df_clean.loc[mask_souche_empty, ["ID", "Souches"]].head(10)

Filas sin especie informativa (Souches vacío): 0 de 9947 (0.00%)


,ID,Souches


In [45]:
# %%
# Alias manuales conocidos (puedes ampliar este diccionario según veas en el value_counts)
SPECIES_ALIAS = {
    "e. coli": "Escherichia coli",
    "e coli": "Escherichia coli",
    "escherichia coli": "Escherichia coli",
    "k. pneumoniae": "Klebsiella pneumoniae",
    "k pneumoniae": "Klebsiella pneumoniae",
    "klebsiella pneumoniae": "Klebsiella pneumoniae",
    "p. mirabilis": "Proteus mirabilis",
    "p mirabilis": "Proteus mirabilis",
    "proeus mirabilis": "Proteus mirabilis",  # typo frecuente
    "citrobacter spp.": "Citrobacter spp.",
    "citrobacter spp": "Citrobacter spp.",
}


def normalize_species(value: object) -> float | str:
    """
    Normaliza la especie:
    - Aplica strip/lower.
    - Usa SPECIES_ALIAS cuando el string coincide.
    - Para el resto:
        * quita dobles espacios,
        * pasa a Title Case,
        * normaliza 'Spp' -> 'spp.'.
    """
    if pd.isna(value):
        return np.nan

    raw = str(value).strip()
    if not raw:
        return np.nan

    lower = raw.casefold()

    # 1) Alias exactos conocidos
    if lower in SPECIES_ALIAS:
        return SPECIES_ALIAS[lower]

    # 2) Normalización genérica
    # quitar id de aislamiento si viene algo tipo "S290 Escherichia coli"
    parts = raw.split(maxsplit=1)
    if len(parts) == 2 and parts[0].startswith(("S", "s")) and parts[0][1:].isdigit():
        raw_name = parts[1]
    else:
        raw_name = raw

    # colapsar espacios
    raw_name = " ".join(raw_name.split())

    # pasar a título
    name_title = raw_name.title()

    # normalizar spp / Spp / spp.
    name_title = name_title.replace("Spp", "spp").replace("Spp.", "spp.")

    return name_title


df_clean["Species_std"] = df_clean["Souches"].apply(normalize_species)

# Vista rápida
df_clean[["Souches", "Species_std"]].head(20)

,Souches,Species_std
0,S290 Escherichia coli,Escherichia Coli
1,S291 Morganella morganii,Morganella Morganii
2,S292 Proteus mirabilis,Proteus Mirabilis
3,S294 Escherichia coli,Escherichia Coli
4,S295 Escherichia coli,Escherichia Coli
5,S296 Escherichia coli,Escherichia Coli
6,S297 Citrobacter spp.,Citrobacter spp.
7,S298 Enterobacteria spp.,Enterobacteria spp.
8,S299 Proeus mirabilis,Proeus Mirabilis
9,S300 Escherichia coli,Escherichia Coli


In [46]:
df_clean

,ID,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt,Age_years,Gender,AMX/AMP_SIR,AMX/AMP_Rbin,AMC_SIR,AMC_Rbin,CZ_SIR,CZ_Rbin,FOX_SIR,FOX_Rbin,CTX/CRO_SIR,CTX/CRO_Rbin,IPM_SIR,IPM_Rbin,GEN_SIR,GEN_Rbin,AN_SIR,AN_Rbin,Acide nalidixique_SIR,Acide nalidixique_Rbin,ofx_SIR,ofx_Rbin,CIP_SIR,CIP_Rbin,C_SIR,C_Rbin,Co-trimoxazole_SIR,Co-trimoxazole_Rbin,Furanes_SIR,Furanes_Rbin,colistine_SIR,colistine_Rbin,Diabetes_bin,Hypertension_bin,Hospital_before_bin,Species_std
0,S290,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05,37.0,Female,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Escherichia Coli
1,S291,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT,29.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0,Morganella Morganii
2,S292,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT,77.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0,Proteus Mirabilis
3,S294,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT,13.0,Female,S,0.0,S,0.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,0.0,1.0,0.0,Escherichia Coli
4,S295,S295 Escherichia coli,True,No,No,3.0,R,S,R,R,S,S,S,R,S,S,S,S,S,S,S,no georeferenciado,NaT,57.0,Male,R,1.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0,Escherichia Coli
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9942,S10994,S8862 Escherichia coli,No,No,No,0.0,S,R,R,R,S,S,S,S,R,S,S,R,S,S,S,RI,NaT,43.0,Female,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Escherichia Coli
9943,S10995,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT,17.0,Male,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Enterobacteria spp.
9944,S10996,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05,20.0,Female,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,1.0,Proeus Mirabilis
9945,S10997,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT,33.0,Male,R,1.0,R,1.0,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Escherichia Coli


In [47]:
# %% [markdown]
# ### Tipificación de Infection_Freq a numérico/ordinal

# %%

INFECTION_FREQ_NA_TOKENS = {"", "?", "none", "missing", "error", "unknown", "na", "n/a"}


def normalize_infection_freq(raw: object):
    """
    Normaliza Infection_Freq a float (0.0-3.0, etc.) o NaN.
    - Tokens basura -> NaN
    - Cadenas numéricas -> float
    """
    if pd.isna(raw):
        return np.nan

    s = str(raw).strip()
    if not s:
        return np.nan

    lower = s.casefold()
    if lower in INFECTION_FREQ_NA_TOKENS:
        return np.nan

    # Intentar parsear a número (puede venir como '0', '1.0', etc.)
    try:
        val = float(s)
        # Filtramos rangos absurdos (si quieres más estricto, ajústalo)
        if -1 <= val <= 10:
            return val
        return np.nan
    except ValueError:
        return np.nan


df_clean["Infection_Freq_num"] = df_clean["Infection_Freq"].apply(
    normalize_infection_freq
)

# Chequeo rápido
n_rows = len(df_clean)
print(df_clean["Infection_Freq_num"].describe())

n_missing_inf = df_clean["Infection_Freq_num"].isna().sum()
print(
    f"\nMissing en Infection_Freq_num: {n_missing_inf} ({n_missing_inf / n_rows * 100:.2f}%)"
)

count    9714.000000
mean        1.514309
std         1.021972
min         0.000000
25%         1.000000
50%         2.000000
75%         2.000000
max         3.000000
Name: Infection_Freq_num, dtype: float64

Missing en Infection_Freq_num: 233 (2.34%)


In [48]:
df_clean

,ID,Souches,Diabetes,Hypertension,Hospital_before,Infection_Freq,AMX/AMP,AMC,CZ,FOX,CTX/CRO,IPM,GEN,AN,Acide nalidixique,ofx,CIP,C,Co-trimoxazole,Furanes,colistine,State,Collection_Date_dt,Age_years,Gender,AMX/AMP_SIR,AMX/AMP_Rbin,AMC_SIR,AMC_Rbin,CZ_SIR,CZ_Rbin,FOX_SIR,FOX_Rbin,CTX/CRO_SIR,CTX/CRO_Rbin,IPM_SIR,IPM_Rbin,GEN_SIR,GEN_Rbin,AN_SIR,AN_Rbin,Acide nalidixique_SIR,Acide nalidixique_Rbin,ofx_SIR,ofx_Rbin,CIP_SIR,CIP_Rbin,C_SIR,C_Rbin,Co-trimoxazole_SIR,Co-trimoxazole_Rbin,Furanes_SIR,Furanes_Rbin,colistine_SIR,colistine_Rbin,Diabetes_bin,Hypertension_bin,Hospital_before_bin,Species_std,Infection_Freq_num
0,S290,S290 Escherichia coli,No,No,No,0.0,R,R,R,R,R,R,S,S,S,S,S,R,S,S,S,RI,2025-02-05,37.0,Female,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Escherichia Coli,0.0
1,S291,S291 Morganella morganii,True,No,No,3.0,S,R,S,R,S,R,S,S,S,S,S,S,S,S,S,HI,NaT,29.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0,Morganella Morganii,3.0
2,S292,S292 Proteus mirabilis,True,No,No,3.0,S,R,S,R,S,S,S,S,S,S,R,R,S,S,S,HI,NaT,77.0,Female,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0,Proteus Mirabilis,3.0
3,S294,S294 Escherichia coli,No,Yes,No,1.0,S,S,R,R,R,R,R,R,S,S,S,S,S,S,S,ME,NaT,13.0,Female,S,0.0,S,0.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,0.0,1.0,0.0,Escherichia Coli,1.0
4,S295,S295 Escherichia coli,True,No,No,3.0,R,S,R,R,S,S,S,R,S,S,S,S,S,S,S,no georeferenciado,NaT,57.0,Male,R,1.0,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,0.0,Escherichia Coli,3.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9942,S10994,S8862 Escherichia coli,No,No,No,0.0,S,R,R,R,S,S,S,S,R,S,S,R,S,S,S,RI,NaT,43.0,Female,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Escherichia Coli,0.0
9943,S10995,S7382 Enterobacteria spp.,No,No,No,3.0,S,R,R,S,S,R,S,S,S,S,S,S,S,S,S,no georeferenciado,NaT,17.0,Male,S,0.0,R,1.0,R,1.0,S,0.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Enterobacteria spp.,3.0
9944,S10996,S1479 Proeus mirabilis,True,No,Yes,3.0,S,R,S,S,S,S,S,S,S,S,S,S,S,S,S,SD,2025-02-05,20.0,Female,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,S,0.0,1.0,0.0,1.0,Proeus Mirabilis,3.0
9945,S10997,S2983 Escherichia coli,No,No,No,0.0,R,R,S,R,R,R,S,S,S,R,S,R,S,S,S,no georeferenciado,NaT,33.0,Male,R,1.0,R,1.0,S,0.0,R,1.0,R,1.0,R,1.0,S,0.0,S,0.0,S,0.0,R,1.0,S,0.0,R,1.0,S,0.0,S,0.0,S,0.0,0.0,0.0,0.0,Escherichia Coli,0.0


In [49]:
# %%
df_clean["Infection_Freq_ord"] = df_clean["Infection_Freq_num"].round().astype("Int64")
df_clean[["Infection_Freq", "Infection_Freq_num", "Infection_Freq_ord"]].head(15)

,Infection_Freq,Infection_Freq_num,Infection_Freq_ord
0,0.0,0.0,0
1,3.0,3.0,3
2,3.0,3.0,3
3,1.0,1.0,1
4,3.0,3.0,3
5,3.0,3.0,3
6,2.0,2.0,2
7,0.0,0.0,0
8,3.0,3.0,3
9,2.0,2.0,2


3️⃣ Cálculo de MDR usando el mapeo antibiótico → clase

3.1. Definir mapeo antibiótico → clase

Basado en la metadata:

* β-lactámicos: AMX/AMP, AMC, CZ, FOX, CTX/CRO, IPM

* Aminoglucósidos: GEN, AN

* Quinolonas/fluoroquinolonas: Acide nalidixique, ofx, CIP

* Otros: C, Co-trimoxazole, Furanes, colistine

In [50]:
# %% [markdown]
# ### Cálculo de MDR (Multidrug Resistance) por clases antibióticas

# %%
ABX_CLASS = {
    "AMX/AMP": "beta_lactams",
    "AMC": "beta_lactams",
    "CZ": "beta_lactams",
    "FOX": "beta_lactams",
    "CTX/CRO": "beta_lactams",
    "IPM": "beta_lactams",
    "GEN": "aminoglycosides",
    "AN": "aminoglycosides",
    "Acide nalidixique": "quinolones",
    "ofx": "quinolones",
    "CIP": "quinolones",
    "C": "other",
    "Co-trimoxazole": "other",
    "Furanes": "other",
    "colistine": "other",
}

CLASS_NAMES = sorted(set(ABX_CLASS.values()))
CLASS_NAMES

['aminoglycosides', 'beta_lactams', 'other', 'quinolones']

3.2. Construir columnas de resistencia por clase

Para cada clase:

Class_[name]_R: 1 si al menos un antibiótico de esa clase está en R (Rbin == 1),
0 si todos los test de esa clase están en S/I (0) o no testados,
NaN si no hay ningún resultado para esa clase en ese aislamiento.

In [51]:
# %%
# Asegurarnos de que existen las columnas *_Rbin
missing_rbin_cols = []
for abx in ANTIBIOTIC_COLS:
    rbin_col = f"{abx}_Rbin"
    if rbin_col not in df_clean.columns:
        missing_rbin_cols.append(rbin_col)

print("Columnas Rbin faltantes:", missing_rbin_cols)

Columnas Rbin faltantes: []


In [52]:
# %%
# Crear columnas de resistencia por clase
class_r_cols = []

for class_name in CLASS_NAMES:
    abx_in_class = [abx for abx, cls in ABX_CLASS.items() if cls == class_name]
    rbin_cols_in_class = [f"{abx}_Rbin" for abx in abx_in_class]

    # Matriz de Rbin de esa clase
    rbin_matrix = df_clean[rbin_cols_in_class]

    # Indica si al menos un antibiótico de la clase fue TESTADO (no todo NaN)
    class_tested = rbin_matrix.notna().any(axis=1)

    # Indica si al menos un antibiótico de la clase está en R (==1)
    class_resistant = (rbin_matrix == 1).any(axis=1)

    # Convertimos a float y ponemos NaN donde la clase no fue testada
    class_resistant = class_resistant.astype("float")
    class_resistant[~class_tested] = np.nan

    col_name = f"Class_{class_name}_R"
    df_clean[col_name] = class_resistant
    class_r_cols.append(col_name)

df_clean[class_r_cols].head(10)

,Class_aminoglycosides_R,Class_beta_lactams_R,Class_other_R,Class_quinolones_R
0,0.0,1.0,1.0,0.0
1,0.0,1.0,0.0,0.0
2,0.0,1.0,1.0,1.0
3,1.0,1.0,0.0,0.0
4,1.0,1.0,0.0,0.0
5,1.0,1.0,0.0,0.0
6,0.0,1.0,1.0,0.0
7,0.0,1.0,0.0,0.0
8,0.0,1.0,0.0,1.0
9,0.0,1.0,0.0,0.0


3.3. Contar clases resistentes y definir MDR

Definimos:

* n_classes_tested: nº de clases con al menos un antibiótico testado.

* n_classes_resistant: nº de clases con al menos un antibiótico en R.

* MDR:

  * 1 si n_classes_resistant >= 3 y n_classes_tested >= 3.

  * 0 si n_classes_resistant < 3 y n_classes_tested >= 3.

  * NaN si n_classes_tested < 3 (insuficiente info para juzgar MDR).

In [53]:
# %%
# Conteo de clases por aislamiento
class_r_df = df_clean[class_r_cols]

# Clase testada = al menos un Rbin no-NaN en la clase
classes_tested = ~class_r_df.isna()
df_clean["n_classes_tested"] = classes_tested.sum(axis=1)

# Clase resistente = valor 1.0 en Class_<name>_R
classes_resistant = class_r_df == 1.0
df_clean["n_classes_resistant"] = classes_resistant.sum(axis=1)


# Definición de MDR
def compute_mdr(row):
    tested = row["n_classes_tested"]
    resistant = row["n_classes_resistant"]
    if tested < 3:
        return np.nan
    return 1.0 if resistant >= 3 else 0.0


df_clean["MDR"] = df_clean.apply(compute_mdr, axis=1)

df_clean[["n_classes_tested", "n_classes_resistant", "MDR"]].head(15)

,n_classes_tested,n_classes_resistant,MDR
0,4,2,0.0
1,4,1,0.0
2,4,3,1.0
3,4,2,0.0
4,4,2,0.0
5,4,2,0.0
6,4,2,0.0
7,4,1,0.0
8,4,2,0.0
9,4,1,0.0


In [54]:
# %%
n_rows = len(df_clean)

print(
    "Distribución de MDR (1 = multirresistente, 0 = no-MDR, NaN = info insuficiente):"
)
display(
    df_clean["MDR"]
    .value_counts(dropna=False)
    .to_frame("count")
    .assign(pct=lambda s: (s["count"] / n_rows * 100).round(2))
)

print("\nResumen de n_classes_resistant:")
print(df_clean["n_classes_resistant"].describe())

Distribución de MDR (1 = multirresistente, 0 = no-MDR, NaN = info insuficiente):


,count,pct
MDR,,
0.0,6863,69.0
1.0,3084,31.0



Resumen de n_classes_resistant:
count    9947.000000
mean        2.071077
std         0.924168
min         0.000000
25%         1.000000
50%         2.000000
75%         3.000000
max         4.000000
Name: n_classes_resistant, dtype: float64


Buenísimo, eso nos dice varias cosas interesantes y además confirma que la lógica de MDR está funcionando como esperábamos 👇

---

## 1. Qué significa ese resultado

De la tabla:

* **0.0 → 6863 aislamientos (69%)**
  → No son multirresistentes según tu definición (resistentes en **≤2 clases** de antibióticos).

* **1.0 → 3084 aislamientos (31%)**
  → **MDR = 1**: resistentes en **≥3 clases** (en tu caso, 3 o 4 clases).

De la descripción de `n_classes_resistant`:

* **media ≈ 2.07 clases resistentes** por aislamiento.
* **P25 = 1**, **P50 = 2**, **P75 = 3**, **máx = 4**.

Eso cuadra perfectamente con tu regla:

```text
MDR = 1 si n_classes_resistant >= 3
MDR = 0 si n_classes_resistant < 3
```

Y como en el conteo de MDR solo aparecen 0 y 1 (sin NaN), eso implica que **para todos los aislamientos había ≥3 clases testadas**, así que nunca entró el caso “información insuficiente”.

---

## 2. Sanity check rápido en el notebook

Si quieres verificar explícitamente que la regla cuadra con los conteos:

```python
# Ver que MDR coincide con el criterio n_classes_resistant >= 3
check = pd.crosstab(df_clean["n_classes_resistant"], df_clean["MDR"])
check
```

Deberías ver algo como:

* Para `n_classes_resistant` = 0, 1, 2 → solo MDR = 0
* Para `n_classes_resistant` = 3 o 4 → solo MDR = 1

---

## 3. Cómo lo podemos usar ya

Con este MDR calculado y revisado, ya podemos empezar a:

* Calcular **prevalencia de MDR por especie** (`Species_std`)

  ```python
  mdr_by_species = (
      df_clean.groupby("Species_std")["MDR"]
      .mean()
      .sort_values(ascending=False)
  )
  ```

* Usar `MDR` como variable dependiente en los modelos con factores de riesgo (`*_bin`, `Age_years`, `Gender`, `State`, etc.).

Si quieres, en el siguiente paso montamos directamente la tabla **MDR por especie** o arrancamos con la **P1: prevalencia especie–antibiótico (%) de R)** usando las columnas `*_Rbin`.


In [55]:
# Ver que MDR coincide con el criterio n_classes_resistant >= 3
check = pd.crosstab(df_clean["n_classes_resistant"], df_clean["MDR"])
check

MDR,0.0,1.0
n_classes_resistant,,
0,222,0
1,2593,0
2,4048,0
3,0,2424
4,0,660


In [56]:
mdr_by_species = (
    df_clean.groupby("Species_std")["MDR"]
    .mean()
    .sort_values(ascending=False)
)
mdr_by_species

Species_std
E.Cli                      0.421717
E.Coi                      0.412037
E. Coli                    0.409207
Escherichia Coli           0.390625
Enteobacteria spp.         0.263158
Pseudomonas Aeruginosa     0.230000
Prot.Eus Mirabilis         0.217391
Proteus Mirabilis          0.217391
Protus Mirabilis           0.215686
Acinetobacter Baumannii    0.215470
Serratia Marcescens        0.191406
Klebsiella Pneumoniae      0.187611
Proeus Mirabilis           0.170213
Morganella Morganii        0.163934
Klebsie.Lla Pneumoniae     0.158730
Enterobacteria spp.        0.147132
Klbsiella Pneumoniae       0.135135
Citrobacter spp.           0.120582
Enter.Bacteria spp.        0.090000
Name: MDR, dtype: float64

Vamos a rehacer Species_std con una función más lista para:

Quitar el prefijo de aislamiento (S290, etc.).

Limpiar puntos y espacios.

Detectar variantes y typos típicos de cada especie
y dejarlas en un único nombre canónico:

Escherichia coli, Klebsiella pneumoniae, Proteus mirabilis,
Pseudomonas aeruginosa, Acinetobacter baumannii,
Serratia marcescens, Citrobacter spp., Enterobacter spp.,
Morganella morganii, etc.

In [65]:
# %% [markdown]
# ### Re-normalización de Species_std (unificar variantes y typos)

# %%


def canonicalize_species(value: object) -> float | str:
    """
    Normaliza el nombre de especie a una forma canónica.
    Unifica variantes tipo 'E.Cli', 'E.Coi', 'E. Coli', 'Escherichia Coli', etc.
    """
    if pd.isna(value):
        return np.nan

    raw = str(value).strip()
    if not raw:
        return np.nan

    # 1) Quitar ID de aislamiento tipo "S290 Escherichia coli"
    parts = raw.split(maxsplit=1)
    if len(parts) == 2 and parts[0].startswith(("S", "s")) and parts[0][1:].isdigit():
        name = parts[1]
    else:
        name = raw

    # 2) Normalización básica
    # - colapsar espacios
    name = " ".join(name.split())
    lower = name.lower()

    # Versión simplificada sin espacios, puntos ni caracteres raros
    simple = re.sub(r"[^a-z]", "", lower)  # solo letras

    # ---- Escherichia coli ----
    if "escherichia" in simple or simple in {"ecli", "ecoi", "ecoli"}:
        return "Escherichia coli"

    # ---- Klebsiella pneumoniae ----
    if (
        "klebsiella" in simple
        or "klebsiel" in simple
        or "klebsll" in simple
        or "klbsiella" in simple
    ) and ("pneum" in simple or "pneumoniae" in simple):
        return "Klebsiella pneumoniae"

    # ---- Proteus mirabilis ----
    if (
        "proteus" in simple
        or "protus" in simple
        or "proeus" in simple
        or "proteusmirabilis" in simple
    ) and "mirab" in simple:
        return "Proteus mirabilis"

    # ---- Pseudomonas aeruginosa ----
    if "pseudomonas" in simple and ("aeruginosa" in simple or "aerug" in simple):
        return "Pseudomonas aeruginosa"

    # ---- Acinetobacter baumannii ----
    if ("acinetobacter" in simple or "acinetob" in simple) and (
        "baumann" in simple or "bauman" in simple
    ):
        return "Acinetobacter baumannii"

    # ---- Serratia marcescens ----
    if "serratia" in simple and ("marcesc" in simple or "marcec" in simple):
        return "Serratia marcescens"

    # ---- Citrobacter spp. ----
    if "citrobacter" in simple:
        return "Citrobacter spp."

    # ---- Enterobacter spp. / Enterobacteria spp. ----
    if (
        "enterobacter" in simple
        or "enterobacteria" in simple
        or "enter.bacteria" in simple
        or "enteobacteria" in simple
        or "enter" in simple
    ):
        return "Enterobacter spp."

    # ---- Morganella morganii ----
    if (
        "morganella" in simple
        or "morganela" in simple
        or "morganellamorganii" in simple
    ):
        return "Morganella morganii"

    # ---- Otros géneros que quizá ya estén bien escritos ----
    # Ponemos en Title Case por defecto (género con mayúscula, resto minúscula)
    # y normalizamos "spp"/"spp."
    name_title = " ".join(w.capitalize() for w in name.split())
    name_title = name_title.replace("Spp", "spp.").replace("Spp.", "spp.")

    return name_title

In [66]:
# %%
df_clean["Species_std"] = df_clean["Souches"].apply(canonicalize_species)
df_clean["Species_std"] = df_clean["Species_std"].astype("category")

# Ver los principales niveles de Species_std
species_counts = df_clean["Species_std"].value_counts().to_frame("count")
species_counts

,count
Species_std,
Escherichia coli,6083
Enterobacter spp.,997
Proteus mirabilis,742
Klebsiella pneumoniae,702
Citrobacter spp.,481
Morganella morganii,305
Serratia marcescens,256
Pseudomonas aeruginosa,200
Acinetobacter baumannii,181


In [68]:
# %%
mdr_by_species = (
    df_clean.groupby("Species_std")["MDR"]
    .mean()
    .sort_values(ascending=False)
    .to_frame("MDR_rate")
)

mdr_by_species.head(20)

C:\Users\samaf\AppData\Local\Temp\ipykernel_18804\4288630082.py:3: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df_clean.groupby("Species_std")["MDR"]


,MDR_rate
Species_std,
Escherichia coli,0.395364
Pseudomonas aeruginosa,0.230000
Acinetobacter baumannii,0.215470
Proteus mirabilis,0.214286
Serratia marcescens,0.191406
Klebsiella pneumoniae,0.179487
Morganella morganii,0.163934
Enterobacter spp.,0.152457
Citrobacter spp.,0.120582
